### ISSUE TITLE
Enhancement: Configuration of different upstream DNS servers for different clients

### ISSUE THREAD

**Issue by adb76** [0xERR0R/blocky#81]
I really like Blocky and the great configuration options. Because it is so slim and fast I changed my LAN setup from Adguard to Blocky. There is only one feature missing in Blocky for me: The possibility to have different upstream DNS servers for different clients. For my understanding it is currently not possible in Blocky to have something like:

```
externalResolvers:  
      default:     # Default for all requests
         - tcp: 46.182.19.48
         - tcp: 80.241.218.68
      desktop.fritz.box:   # Requests coming from desktop.fritz.box will be forwarded to
          - tcp-tls:fdns1.dismail.de:853
          - https://dns.digitale-gesellschaft.ch/dns-query
```

Or maybe I missed something in the configuration options?

**Comment by 0xERR0R** [comment]
This is correct, currently, there is no configuration option to achieve this. Could you describe your use case please? Maybe there exist some other options?

**Comment by adb76** [comment]
The idea behind the use case is:

- I want to use the default upstream DNS Server from my internet provider for most of my devices
- For some devices (e.g. video streaming devices) I want to set a special DNS to solve geo location issues ;-)
- I don't won't to set the special DNS server to all of my devices, because of latency etc.
- Of course it is possible to set that DNS server in the configuration of the client, but it is much more comfortable to do that centrally, the way for example AdGuard offers it.

**Comment by 0xERR0R** [comment]
#43

**Comment by Jogai** [comment]
It it something you're considering adding? Where should we vote? Here of in #43?

**Comment by 0xERR0R** [comment]
Yes, this feature is useful and I'll implement it in the (near) future. I would like to avoid breaking changes in the configuration (like example in the description -> this will break old installations), so maybe new configuration option is needed. You can vote here, I'll close #43

**Comment by Jogai** [comment]
Sounds good! Maybe follow what you did for the client groups for blocking, then you'd end up with a config like this, which wouldnt break an existing config:
```
upstream:
  externalResolvers:
    - 1.1.1.3
    - 1.0.0.3
  malwareBlockingResolvers:
    - 1.1.1.2
    - 1.0.0.2
resolvers:
    default:
      - externalResolvers
    nsfw*:
      - malwareBlockingResolvers
```

**Comment by 0xERR0R** [comment]
I think, grouping upstream resolvers is a bit over-complicated. In case of blocking goups -> it is useful, since your client can be in several groups. For upstream resolvers there should be always 1:1 mapping.

I think, we can start with something like this:

```
upstream:
  externalResolvers:
    - 46.182.19.48
    - 1.1.1.1
  laptop*:
    - 8.8.8.8
    - 9.9.9.9
```

`externalResolvers` or `default`ist the default fallback configuration. Otherwise using the name (like "laptop*" here as example) to match agains the client name. Can also be IP or CIDR.

What do you think?

**Comment by Jogai** [comment]
Looks good, and indeed, even simpler to apply! 

For me it would be enough, but what if you have devices with wildly different names? Don't you still need client groups?

**Comment by tahngarth** [comment]
The fact that it can be a CIDR makes this work perfectly for my use case. For Jogai's concern could you also make it either check for an exact client group name, or maybe some sort of prefix sigil to mark that it's a client group?

**Comment by 0xERR0R** [comment]
This will be the same logic as for blocking groups (detailed description [here](https://0xerr0r.github.io/blocky/configuration/#client-groups) You can define:

- client name exact
- client name with wildcards (file pattern logic): example: `laptop*` or `*workstation[1-3]`
- client's IP
- CIDR

**Comment by 0xERR0R** [comment]
Resolved by #193
